File: Exercise_1.cpp
/* 
    Time Complexity                              :  O(N) - fixed time
    Space Complexity                             :  O(1)
    Did this code successfully run on Leetcode   :  Yes
    Any problem you faced while coding this      :  No
*/

#include <bits/stdc++.h> 
using namespace std;  

https://leetcode.com/problems/array-partition-i/

class Solution {
public:
    int arrayPairSum(vector<int>& nums) {
        unordered_map<int,int> mp;
        for(auto num : nums) {
            mp[num]++;
        }
        int f = 0, ans =0;
        for(int i=-10000;i<=10000;) {
            if(mp[i] > 0 and f == 0) {
                ans += i;
                mp[i]--;
                f = 1;
            } else if (mp[i] > 0 and f == 1) {
                mp[i]--;
                f = 0;
            } else {
                i++;
            }
        }
        return ans;
    }
};

File: Exercise_1.py
# Problem1 Array partition (https://leetcode.com/problems/array-partition/)
# // Time Complexity : O(nlogn)
# // Space Complexity :O(1)
# // Did this code successfully run on Leetcode :Yes
# // Any problem you faced while coding this :No
class Solution:
    def arrayPairSum(self, nums):
        # sorting the nums
        nums = sorted(nums)
        n = len(nums)
        # starting from back
        i = n - 1
        # variable for counting total to 0
        curr_sum = 0
        while i >= 0:
            curr_sum += min(nums[i], nums[i - 1])
            i -= 2
        return curr_sum


nums = [1, 4, 3, 2]
sol = Solution()
print(sol.arrayPairSum(nums))
